### PR TITLE
Introduce ToggleColumn for Grid component

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop |
| Description?  | Introduce a new type of Column for Grid component |
| Type?         | new feature |
| Category?     | BO |
| BC breaks?    | no |
| Deprecations? | no |
| Fixed ticket? | 
| How to test?  | See below

## About the feature

This introduces the `PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\ToggleColumn` class with its Twig template and the JS assets to allow its behavior.

This Column is used to display booleans.
- it will display an icon instead of the value
- if user clicks on it, this triggers a toggle of the boolean value

In the current implementation, when the user clicks on a togglable row, the JS asset sends a POST request to `/toggle/{row_id}`.

<br/>

This means that on the page ```http://localhost/admin-dev/index.php/configure/advanced/logs/?_token=K0PaeKcR6PF5HDs3HqjmGFLEZ0kmlJTvgydEcq6ZiSc```

the POST request is sent to URL ```http://localhost/admin-dev/index.php/configure/advanced/logs/toggle/4?_token=K0PaeKcR6PF5HDs3HqjmGFLEZ0kmlJTvgydEcq6ZiSc``` for example.

⬇️ **is that OK @sarjon @mickaelandrieu ?**

### Usage

In the DefinitionFactory of the grid:
```
    /**
     * {@inheritdoc}
     */
    protected function getColumns()
    {
        return (new ColumnCollection())
            ->add(...
            )
            ....
            ->add(
                (new ToggleColumn('is_enabled'))
                ->setName($this->trans('Enabled', [], 'Admin.Advparameters.Feature'))
                ->setOptions([
                    'field_to_display' => 'is_enabled',
                    'field_to_use_as_id' => 'id',
                ])
            )
```

### Requirement: toggle action

This requires to setup a Controller action to perform the toggle:

*Routing*
```
admin_toggle:
    path: /toggle/{id}
    methods: [POST]
    defaults:
        _controller: 'PrestaShopBundle:Admin\Configure\AdvancedParameters\Logs:toggle'
        _legacy_controller: AdminLogs
```

*Controller*
```
    /**
     * @AdminSecurity("is_granted('delete', request.get('_legacy_controller')~'_')", message="You do not have permission to update this.", redirectRoute="admin_listing")
     *
     * @return RedirectResponse
     * @throws \Doctrine\DBAL\Exception\InvalidArgumentException
     */
    public function toggleAction($id)
    {
        // perform toggle of the boolean value for $id     
    }
```

## How to test

Cannot be tested by QA, developers only. Validation of the "Advanced Parameters > Webservice" page later will validate this feature.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9305)

### CODE REVIEW

**Comment by matks** [comment]
This is what it looks like:
capture d ecran 2018-07-11 a 12 42 36

The "disabled" icon is not very good 😢 but I could not find a cross icon in Material Icons https://material.io/tools/icons/

**Comment by matks** [comment]
This is what it looks like in legacy page
capture d ecran 2018-07-11 a 15 27 56

**Comment by sarjon** [comment]
check if you can take icons from catalog page. https://prnt.sc/k56x33

**Comment by mickaelandrieu** [comment]
> Requirement: toggle action

I'm not easy with it, we must rely on a Common Controller action, with a Hook to allow people from modules to decide what needs to be done on toggle action.

Introducing it this way, you break the expected features and behaviors of the component:

* Hooks to manipulate the data
* Templates overrides to manipulate the view

There won't be anything more accepted in this component to enrich the grid :no_entry: 

Also, I've noticed you always put "SF Grid". This component is not related to Symfony at all, it's a new and - for now - totaly uncoupled to PrestaShop component. The way I see it, we could totally imagine to extract it later, and provide a package and a bridge for Twig.

The only issue I see is the use of HookDispatcher but this is something we will improve later (on my giant todo list named "Save the world, improve PrestaShop Code").

File: admin-dev/themes/new-theme/js/components/grid.js
Comment (sarjon) [inline admin-dev/themes/new-theme/js/components/grid.js]: it keeps getting bigger and bigger. what if we could do something like this:

```javascript
// column-toggle-handler.js

class ColumnToggleHandler() {
    run(event, grid) {
        // do stuff
    }
}

// some/specific/page/script.js

const grid = new Grid('#my-grid');

// you just add handlers that you actually need
// it can be the same for column, grid actions & etc.
// maybe fire some event so modules/other script can register their own handlers for the grid
grid.addColumnHandler(new ColumnToggleHandler());

grid.init();
```

@mickaelandrieu what do you think?

**Comment by matks** [comment]
@mickaelandrieu OK, so if I get correctly:
- the POST request on toggling should target an action from `PrestaShopBundle\Controller\Admin\CommonController`
- the action should perform a `dispatchHook()`, nothing else
- when I migrate the Webservice Controller, I should listen on this hook and catch the dispatch to perform the database update

Is that right ?

File: admin-dev/themes/new-theme/js/app/utils/column-toggling.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/column-toggling.js]: If parameter is a jQuery table why reusing `$()`?

File: admin-dev/themes/new-theme/js/app/utils/column-toggling.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/column-toggling.js]: No need to use const $row

File: admin-dev/themes/new-theme/js/app/utils/column-toggling.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/column-toggling.js]: You can use es6 syntax  (and it's better)
```js
$.post(`toggle${data('toggleFieldId')}?_token${url.searchParams.get('_token')}`).done(() =>
  location.reload();
}).fail(() => {
  console.error(`Failed to toggle row value ${row.data('toggleFieldId')) row.data('toggleFieldName')`);
});
```

File: src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/toggle.html.twig
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/toggle.html.twig]: Weird indent

File: admin-dev/themes/new-theme/js/components/grid.js
Comment (mickaelandrieu) [reply admin-dev/themes/new-theme/js/components/grid.js]: yes, but in a specific Pull request pls :)

Can you think about a system when module developers can add more handlers without touch grid.js file, please? Create the issue on `prestafony-project`, you have already work to do today, isn't it?

File: admin-dev/themes/new-theme/js/app/utils/column-toggling.js
Comment (mickaelandrieu) [reply admin-dev/themes/new-theme/js/app/utils/column-toggling.js]: This is not how I want people to share uris/urls in this component, see how you can do it => https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Definition/Factory/LogGridDefinitionFactory.php#L150

To sum up: we never build urls/uris in Javascript.

**Comment by mickaelandrieu** [comment]
> OK, so if I get correctly

yes.

Send data in POST, you can rely on what is done in `reset_search` and `CommonController::resetAction()`, 

In the hook, pass the data you need: 'item_id' for instance.

Regarding your current action: we need to call both the hookDispatcher and the EventListener :)
So you may listen, the event and do your action at the right place.

**Comment by mickaelandrieu** [comment]
@matks thinking about it, I'm not sure we should do that (hook + dispatch an event).

From our side (core migration) I'm ok to use toggleAction by convention, but the ToggleColumn should rely only on urls to push in POST to the server: it's up to developers to decide which url should be called.

So a PrestaShop developer may call a Module controller to do what he needs to do on toggle action: wdyt?

From our side (the core devs) we will use toggleAction everywhere (and we may probably find a way to abstract stuff if code allow it to do ^^).

**Comment by matks** [comment]
@mickaelandrieu
Indeed It makes it more extensible to allow developers to choose the URL to push to. So that is a good idea.

However then comes the issue of the route arguments.
In a "classic" usecase we might configure a ToggleColumn to send POST requests to `/admin-dev/common/toggle/{controller_name}/{row_id}`.

However what if the module developer wants to send POST request to `/my-url/toggle/{field_one}/{row_id}?toggleOnlyIfValid={is_valid}` ?

To allow developers to pass any fields from the `$row` object to the ToggleColumn we might have to do something like this:
```
            ->add(
                (new ToggleColumn('is_valid'))
                ->setName($this->trans('is_valid', [], 'Global.Actions'))
                ->setOptions([
                    'field' => 'id', // property to use as an ID for the row
                    'route' => 'my_route', // SF route name
                    'arguments' => $myMapping
                ])
            )
```

And `$myMapping` would be a object which says "you need to extract properties foo1, foo2, foo3 from the row pass them to the router to build the URL". In my previous example the `$myMapping` object would allow the ToggleColumn to find out what are the `field_one`, `row_id` and `is_valid` parameters needed by the route.

By the way the same issues applies to Row Actions such as "Edit this row" or "Delete this row". In a simple scenario, you only need the row ID to build the URL. But in a complex scenario you have to provide a mapping for the route parameters.
I prototyped it in https://github.com/PrestaShop/PrestaShop/pull/9272/files#diff-3cc76c33621e62d769f79f4db4515a78R116 with an array instead of a mapping object but I think an object is more reliable.

It works but it is not simple. Is it too much . I guess we want to make things as simple as possible for PS developers. 😢

**Comment by mickaelandrieu** [comment]
Yes, thinking of it "our" ToggleColumn should only manage our use case. If a developer needs more complex use case... he will implement his own Column!

Let's rely on an url only, wdyt?

**Comment by matks** [comment]
@mickaelandrieu It makes sense. I'm starting the implementation right now.

**Comment by matks** [comment]
I fixed my icons design (right icons + right colors). In order to do that I needed to introduce a new CSS class. I looked at what already exists and I did what seemed right.

Hope it was the right place / the right way

https://github.com/PrestaShop/PrestaShop/pull/9305/commits/392945fa62b85116688a0977378e9b58735fdfc8

File: src/Core/Grid/Column/Type/Common/ToggleColumn.php
Comment (sarjon) [inline src/Core/Grid/Column/Type/Common/ToggleColumn.php]: maybe we can use `field` name? so we are consistent with `DataColumn`, what do you think? :)

File: src/Core/Grid/Column/Type/Common/ToggleColumn.php
Comment (sarjon) [inline src/Core/Grid/Column/Type/Common/ToggleColumn.php]: maybe `field_primary` or `toggle_field` would be better?

File: src/Core/Grid/Column/Type/Common/ToggleColumn.php
Comment (sarjon) [inline src/Core/Grid/Column/Type/Common/ToggleColumn.php]: this one can also be simplified, maybe `toggle_url`? :) 

p.s. can it be route name instead? so you can generate url in template. :)

Comment (mickaelandrieu) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php]: even easier => url

Comment (matks) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php]: @sarjon I can dot it but it would not be consistent with other [LogGridDefinitionFactory columns](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Definition/Factory/LogGridDefinitionFactory.php#L50). Is that OK ?

Comment (mickaelandrieu) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php]: this doesn't have to be as GridDefinition factories and Column are different things, you may compare with others Column types. As naming is quite hard, let's give us time for now with the chose to update it until the release of 1.7.5, ok? :)

Comment (matks) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php]: OK but I need it to migrate page "Configure > Advanced Parameters > Webservices" grid 😉

Comment (mickaelandrieu) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php]: It's not blocking, we can merge it with this naming already :+1:

**Comment by matks** [comment]
@mickaelandrieu PR rebased and updated to comply with latest refacto of the grid component

File: src/Core/Grid/Column/Type/Common/ToggleColumn.php
Comment (sarjon) [inline src/Core/Grid/Column/Type/Common/ToggleColumn.php:56]: i would prefer having `route` so you can generate it in template. This would avoid injecting generated route or Router itself into definition factory.

Comment (mickaelandrieu) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php:56]: this could be improved by tomas, the idea is to get the code fast :)

Comment (sarjon) [reply src/Core/Grid/Column/Type/Common/ToggleColumn.php:56]: no problem. :)

**Comment by matks** [comment]
PR rebased (for assets) and updated to comply with latest code-style requirements

**Comment by mickaelandrieu** [comment]
Believe me or not, but I'll merge this one if the build is green :angel:

**Comment by matks** [comment]
@mickaelandrieu Build failed because `apt-get install` failed 😭 
I restarted the build

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Thanks @matks and everyone for review

**Pull request merged**
